### PR TITLE
oo-ride-share - Liz and Hannah

### PR DESCRIPTION
**Pull request by hertweckhr1**
# OO Ride Share
Congratulations! You're submitting your assignment!

## Comprehension Questions
Question | Answer
:------------- | :-------------
Describe a design decision you had to make when working on this project. What options were you considering? What helped you make your final decision? | When we were working on wave 2, we were originally trying to make our code align with the spec files.  We came to the decision to end up changing some of the spec files to fit the requirements, because the spec files were not aligned with changes to the code requirements. We were constantly hit with errors, and realized this was the right to make. 
Describe any examples of composition that you encountered in this project, if any | We had a user, class, driver class, and trip dispatcher.  A passenger and driver can both have many trips.  A trip can only have one passenger and one driver.  
Describe the relationship between `User` and `Driver`|  Driver is a subclass of User.  It inherits attributes from the user class, while also adding attributes of its own like driven_trips and status.  
Describe a _nominal test_ that you wrote for this assignment. | We wrote a test to test if it assigns an available driver.  It checks to make sure in the event of requesting a trip, the driver id must equal the expected value from the test file. 
Describe an _edge case test_ that you wrote for this assignment | An edge case that we wrote, is if there are no available drivers, we then add an argument error. 
Describe a concept that you/your pair gained more clarity on as you worked on this assignment |
What are two discussion points that you and your pair discussed when giving/receiving feedback from each other that you would be willing to share? |  We realized that the code we're given may not be in perfect form, so it's wrong to make the assumption it is. Through that we learned more about deciphering errors, and editing spec files to help code run successfully.  We also learned more about the use of inheritance and object oriented programming.  As far as pair feedback - we both thought we knew when to go and ask when we did not understand.  On a way to improve, we often had good ideas, but instead with trying one way, we ended up trying many different variations at the same time, and then code often became very confusing.  In the future we think it would be beneficial to focus on one idea at a time.

### CODE REVIEW

File: lib/trip_dispatcher.rb [hunk 1]
```diff
@@ -67,6 +75,63 @@ def inspect
...
+    def find_driver(id)
+      check_id(id)
+      return @drivers.find { |driver| driver.id == id }
+    end
+
+    def request_trip(user_id)
+      passenger = find_passenger(user_id)
+
+      available_drivers = @drivers.find_all { |driver| driver.status == :AVAILABLE && driver.id != user_id}
+         raise ArgumentError, 'No available drivers' if available_drivers == []
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:111]: Minor issue: Shouldn't be indented

File: lib/driver.rb [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+    def add_driven_trip(trip)
+      unless trip.instance_of? Trip
+        raise ArgumentError, 'invalid trip data'
+      end
+      @driven_trips << trip
+    end
+
+    def add_trip_in_progress(trip)
+      @status = :UNAVAILABLE
+      @driven_trips << trip
```
Comment (CheezItMan) [inline lib/driver.rb:47]: Why not call `add_driven_trip` instead?

File: lib/driver.rb [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+      completed_trips = @driven_trips.find_all { |trip| trip.end_time != nil }
+      total = 0
+      completed_trips.each do |trip|
+        total += (trip.cost - 1.65) * 0.80
+      end
+      return total
+    end
+
+    def net_expenditures
+      return (super - total_revenue)
```
Comment (CheezItMan) [inline lib/driver.rb:60]: 👍

File: lib/driver.rb [hunk 4]
```diff
@@ -0,0 +1,70 @@
...
+    end
+
+# latest drive developed for wave 4
+    def latest_drive
```
Comment (CheezItMan) [inline lib/driver.rb:64]: Very clever method, you're not assuming the most recent trip is the last one in the array.  That's good!

File: lib/trip_dispatcher.rb [hunk 5]
```diff
@@ -67,6 +75,63 @@ def inspect
...
+      end
+      return drivers
+    end
+
+    def find_driver(id)
+      check_id(id)
+      return @drivers.find { |driver| driver.id == id }
+    end
+
+    def request_trip(user_id)
```
Comment (CheezItMan) [inline lib/trip_dispatcher.rb:107]: This method is well done, but it's a bit long.  You could break it into some helper methods doing the tasks of:

1.  Selecting an available driver
1.  Creating a new trip

File: specs/user_spec.rb [hunk 6]
```diff
@@ -58,5 +72,39 @@
         expect(trip.passenger.id).must_equal 9
       end
     end
+
+    it 'calculates the net expenditure of all trips for the user' do
```
Comment (CheezItMan) [inline specs/user_spec.rb:76]: You should also test this with a person with no trips.

**Comment by CheezItMan** [comment]
# Ride Share
## What We're Looking For

Feature|Feedback
:------------- | :-------------
**Baseline** |
Used Git Regularly | Check, good number of commits and good commit messages, you used branches too!
Answer comprehension questions | Check
**Wave 1** |
Appropriate use of Ruby's `Time`| Check
`Trip` has a helper method to calculate duration|  Check
`User` (passenger) has a method to calculate total cost of all trips| Check
Tests for wave 1 | Check
**Wave 2** |
`Driver` inherits from `User` | Check
`Driver` has `add_driven_trip` method | Check
`Driver` has method to calculate average rating | Check
`Driver` has method to calculate net expenditures and it uses `super` | Check
`Driver` has a method to calculate total revenue | Check
Tests for wave 2 | 
**Wave 3** |
`TripDispatcher` has a new method to create trips | Check
creating a trip in `TripDispatcher` relies on methods in `Driver` and `User` (passenger) to modify their own attributes | Check
Complex logic was correctly implemented | Check
Tests for `request_trip` | Check
Methods from wave 1 and 2 handle incomplete trips | Check
Tests for wave 1 and 2 methods with incomplete trips | Check
**Wave 4 (Optional)** |
`TripDispatcher` now assigns trips to either the newest driver (no trips), or the driver who has not driven in the longest time | Good logic for determining the driver.  You are selecting a driver with no trips or a driver with the oldest trip.
Appropriate helper methods were made to help with complex logic | Yes, although `request_trip` could be broken up further.  You do have a great helper method in `Driver`.  
Tests for wave 4 | Good tests written for `request_trip`.  
**Overall** | You did really well.  You have some great tests and solid application code.  Great work!
